### PR TITLE
soc: esp32/s2/c3: drivers: counter: RTC timer based implementation

### PR DESCRIPTION
**Pull request by glaubermaroto**
Add the counter driver API implementation based on Espressif's RTC timer.

### CODE REVIEW

File: dts/riscv/espressif/esp32c3.dtsi [hunk 1]
```diff
@@ -71,7 +71,16 @@
 			label = "RTC";
 			xtal-freq = <ESP32_CLK_XTAL_40M>;
 			#clock-cells = <1>;
-			status = "ok";
+			status = "okay";
+
+			rtc_timer: rtc_timer {
+				compatible = "espressif,esp32-rtc-timer";
+				label = "RTC_TIMER";
```
Comment (galak) [inline dts/riscv/espressif/esp32c3.dtsi:78]: remove `label` property.

File: dts/xtensa/espressif/esp32.dtsi [hunk 2]
```diff
@@ -67,7 +67,16 @@
 			label = "RTC";
 			xtal-freq = <ESP32_CLK_XTAL_40M>;
 			#clock-cells = <1>;
-			status = "ok";
+			status = "okay";
+
+			rtc_timer: rtc_timer {
+				compatible = "espressif,esp32-rtc-timer";
+				label = "RTC_TIMER";
```
Comment (galak) [inline dts/xtensa/espressif/esp32.dtsi:74]: remove `label` property.

File: dts/xtensa/espressif/esp32s2.dtsi [hunk 3]
```diff
@@ -67,7 +67,16 @@
 			label = "RTC";
 			xtal-freq = <ESP32_CLK_XTAL_40M>;
 			#clock-cells = <1>;
-			status = "ok";
+			status = "okay";
+
+			rtc_timer: rtc_timer {
+				compatible = "espressif,esp32-rtc-timer";
+				label = "RTC_TIMER";
```
Comment (galak) [inline dts/xtensa/espressif/esp32s2.dtsi:74]: remove `label` property.

File: dts/bindings/counter/espressif,esp32-rtc-timer.yaml [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        such situations.
+
+        There is also no need to enable the RTC Timer node, it starts
+        running from power-up.
+
+include: base.yaml
+
+properties:
+    label:
+      required: true
```
Comment (galak) [inline dts/bindings/counter/espressif,esp32-rtc-timer.yaml:20]: remove `label`

**Review by galak — DISMISSED** [review]
remove label property from dts binding and .dtsi files.

File: drivers/counter/counter_esp32_rtc.c [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+
+static int counter_esp32_get_value(const struct device *dev, uint32_t *ticks)
+{
+	ARG_UNUSED(dev);
+
+	SET_PERI_REG_MASK(RTC_CNTL_TIME_UPDATE_REG, RTC_CNTL_TIME_UPDATE);
+#if defined(CONFIG_SOC_ESP32)
+	while (GET_PERI_REG_MASK(RTC_CNTL_TIME_UPDATE_REG, RTC_CNTL_TIME_VALID) == 0) {
+		/* might take 1 RTC slowclk period, don't flood RTC bus */
+		k_sleep(K_USEC(1));
```
Comment (uLipe) [inline drivers/counter/counter_esp32_rtc.c:92]: Why not use busy_wait here?

**Review by uLipe — APPROVED** [review]

File: drivers/counter/counter_esp32_rtc.c (hunk 5 above)
Comment (glaubermaroto) [reply drivers/counter/counter_esp32_rtc.c:92]: It may be a matter of personal taste, but the intent looks clearer to me this way.

**Review by sylvioalves — APPROVED** [review]

**Comment by sam131208** [comment]
@glaubermaroto 
I noticed that the clock of the RTC is not using XTAL 32.768Khz, can the RTC be configured to this clock?

**Pull request merged**